### ISSUE TITLE
The Steam client loses non-default game folder settings on abnormal shutdown

### ISSUE THREAD

**Issue by ezdev128** [ValveSoftware/steam-for-linux#8552]
#### Your system information

* Steam client version (build number or date): 1647446817
* Distribution (e.g. Ubuntu): Ubuntu
* Opted into Steam client beta?: [Yes/No] No
* Have you checked for system updates?: [Yes/No] Yes

#### Please describe your issue in as much detail as possible:
The Steam client loses non-default game folder settings on abnormal shutdown. 
After restarting the Steam client, you have to re-attach the alternative game path. 
In this case, no other settings are broken. 
This has been going on for half a year.

#### Steps for reproducing this issue:

Any of:
1. after execution kill command on any hanged game while the Steam client is running
2. after execution kill command on Steam client while any game is running
3. right click on tray Steam icon -> 'Exit from steam' while any game is running

**Comment by kisak-valve** [comment]
Hello @ezdev128, can you clarify what you mean by "non-default game folder settings." Where is it configured? This doesn't sound like something the Steam client has ever supported.

**Comment by ezdev128** [comment]
@kisak-valve I mean 'Steam Library Folders'.
Default is: /home/$USER/.steam/debian-installation.
So any else game folders (except default) in Storage Manager are affected.

**Comment by apcolvin** [comment]
I have also had this happen to me several times and I use opensuse on two separate machines and the same on both - never found the trigger though.  An additional piece of info for me is that my secondary location is /games/steam/....  and the library folders lists the location in the dropdown as /games only.  Additionally when I reconnect the library all of the games shaders in that library get redownloaded.  Hope that helps track it down

**Comment by kisak-valve** [comment]
Hello @ezdev128, what is the fill path you used with your secondary Steam library folder(s)? I suspect that there's a pattern here of Steam libraries with paths ending in `[...]/steam`.

Another example seen in #8596 with the Steam library located at `/mnt/arch-games/steam`.

**Comment by apcolvin** [comment]
Mine ends /steam as you suggest with games in steam/steamappsOn 18 May 2022 13:34, kisak-valve ***@***.***> wrote:
Hello @ezdev128, what is the full path you used with your secondary Steam library folder(s)? I suspect that there's a pattern here of Steam libraries with paths ending in [...]/steam.
Another example seen in #8596 with the Steam library located at /mnt/arch-games/steam.

**Comment by TTimo** [comment]
Hello @ezdev128,

We cannot reproduce this. Can you provide the following files maybe: `config/libraryfolders.vdf` `steamapps/libraryfolders.vdf` and `logs/content_log.txt` (under `~/.steam/steam` if you use the official Valve install paths).

If the `libraryfolders.vdf` files are still pointing to your external steam library once the client has stopped loading it, consider archiving it and uploading it to us (you can probably remove common/ which is going to be the bulk of the content, or reproduce on an empty library).

**Comment by asimshrestha2** [comment]
This has been happening to me recently, usually after a reboot from a long session use.

Here are the files that you mentioned: [gist link](https://gist.github.com/asimshrestha2/0528fff541562e653295c380f1f39822)

Also noticed that it says `[2022-09-21 00:24:47] Drive removed: /mnt/NVMEData` but I never did that

I am using Pop OS 22.04, installed using the distro-provided steam package, and using the Steam Client Beta version 1663634272

**Comment by ashtonx** [comment]
Have similar issue on arch. Rather than library, it forgets the games in it, also since a while I can't select other library as default  one, this is real pain since my original ssd is only 128gb.

**Comment by arochfold** [comment]
I had the issue on Ubuntu that Steam kept forgetting my manually-added **default** library folder. I believe it's solved (survived 2 reboots...) by changing the mount point to /home/xyz rather than /mnt/ - as recommended here https://support.system76.com/articles/extra-drive/.

**Comment by ashtonx** [comment]
> I had the issue on Ubuntu that Steam kept forgetting my manually-added **default** library folder. I believe it's solved (survived 2 reboots...) by changing the mount point to /home/xyz rather than /mnt/ - as recommended here https://support.system76.com/articles/extra-drive/.

Sadly no, it's still an issue, it seems like sometimes steam doesn't close properly when you're pressing shut down button while game is running and loses the libraries and you need to re-add them.

Try doing shutdown while game is running few times and you'll most likely encounter it again.

**Comment by zany130** [comment]
same issue here [steam sys info](https://gist.github.com/zany130/6cf95d6a172b476b1e1a93d0be5b3485)

**Comment by kermitfrog** [comment]
I lately had problems with this, too.
System is a current KDE Neon (Ubuntu 22.04 LTS based)
I'm not sure if this only happens on an abnormal shutdown, but keep a closer look on shutdown conditions.
My secondary paths end in SteamSSD and SteamLibrary (which than contain libraryfolder.vdf and steamapps)

**Comment by ltallen** [comment]
I believe I am experiencing this as well. I never made a connection with abnormal shutdowns, but I do fairly often have to kill Hearts of Iron 4 when it freezes, so they could be connected.

**Comment by Trollman16747** [comment]
Affects me as well, steam forgets other libraries, happens when steam gets closed/crashes while game is running. Running steam in archlinux via distrobox under fedora kinoite host.

**Comment by NathanielH-snek** [comment]
Not to necro this thread, but I am also having this issue on Ubuntu. Seems this hasn't been fixed yet

**Comment by TTimo** [comment]
Next beta will have some changes that may improve this issue.

**Comment by TTimo** [comment]
Latest beta should have fixed this, if it hasn't please provide `content_log.txt` next time it happens (ttimo@valvesoftware.com) as it'll have more diagnostic information.

**Comment by zany130** [comment]
still not working for me on latest beta 
[content_log.txt](https://github.com/ValveSoftware/steam-for-linux/files/11410870/content_log.txt)

**Comment by Kyr4l** [comment]
glad to see i'm not the only once facing this problem
tho i don't think rebooting via 'sudo reboot now' is what i'd call an abnormal shutdown

**Comment by zany130** [comment]
happened again here is another content log this time I will attach both `content_log.previous.txt` and `content_log.txt`
[content_log.previous.txt](https://github.com/ValveSoftware/steam-for-linux/files/11657863/content_log.previous.txt)
[content_log.txt](https://github.com/ValveSoftware/steam-for-linux/files/11657864/content_log.txt)

**Comment by TTimo** [comment]
@zany130 what is `/mnt/GAMES` on your system? What filesystem? Is it a removable device?

The relevant lines:
```
[2023-06-01 22:22:38] Drive removed: /mnt/GAMES
[..]
[2023-06-01 22:22:38] Unmounted library folder "/mnt/GAMES/SteamLibrary" (68 apps)
[2023-06-01 22:22:38] Triggered async write of Steam library folders configuration (1 libraries).
```

Steam gets a notification from the OS that the partition was unmounted, and so it removes the library. The client doesn't keep anything around, since the filesystem no longer exists. The library and the games are gone.

I'll have to check with other devs here but I'm not sure that's 'wrong'. On Deck there's code to detect and mount the library again if an sd card gets inserted, but there's nothing like that on desktop. Maybe the Steam client should keep track of UUIDs it has seen before and match them up to try and re-add a library.

It seems like for now the OS shouldn't unmount the partition from under Steam's nose.

**Comment by zany130** [comment]
Yes, that's my secondary Steam library on a SATA hard drive I have permanently connected.

It should be mentioned it's not being removed consistently. So sometimes I forcibly kill steamn (or cut the power to my PC because it hangs) then re-open Steam, and it doesn't always remove the drive (before it was)

**Comment by JarnoKai** [comment]
I'm facing this issue on a fresh install of Ubuntu 22.04.2.
Looking at logs/content_log.txt it seems obvious why it's happening:
```
[2023-06-20 23:54:02] Drive removed: /mnt/Jättis
[2023-06-20 23:54:02] Dependency removed: parent 228980, child 12120, depot 228990
... dependency messages for about 50 rows or so
[2023-06-20 23:54:02] Dependency removed: parent 228980, child 2090230, depot 228990
[2023-06-20 23:54:02] Unmounted library folder "/mnt/Jättis/SteamLibrary" (56 apps)
```
So, when I shut down my system, the drive gets unmounted by the OS before steam is shut down. Steam is notified and it removes the library folder. Obviously this is intended behaviour, but I wonder what one could do to work around it? Obviously shutting down steam first and then the OS will help but a less manual workaround would be nice.
Worth noting that I didn't have this issue on my previous install. The library folder is the same and even the /etc/fstab row for the drive is copied over from the old installation.

**Comment by TTimo** [comment]
Thanks @JarnoKai. We have two confirmations of the root cause for this now.

This is really an OS problem - Ubuntu 22 only as far as I can tell? UI / desktop sessions processes shouldn't be left running while the OS is unmounting partitions...

We could put a workaround in the Steam client, either by detecting that the system is shutting down, or by keeping a mapping of library paths indexed by partition UUIDs so we can remount them later. I can't give you an ETA on us doing this though.

**Comment by uqs** [comment]
This just hit me, though I'm not sure I can reconstruct what happened. Was playing some game on Arch ... stopped that, but Steam was still running. Did a system update (including Kernel) and attached/detached some USB sticks.

Then (w/o rebooting) wanted to start Hunt: Showdown, it ran but caused some exception ... I figured I should probably reboot to get everything aligned, did that, Steam library is empty :/

Under Downloads, I only see `/home/uqs/.local/share/Steam` and my other location of /opt/steam is gone. NB: /opt is _not_ a separate filesystem, but resides on / just as /home, so it's essentially a symlink-like situation.

With the UI bug in the new steam (see issue #9273 ) I also can't seem to re-add /opt/steam ... I'll probably symlink things over at this point to get things unstuck?

**Comment by uqs** [comment]
Ok, lol. So I had a look at `.local/share/Steam/steamapps/libraryfolders.vdf` and it looked hopeless, but then I saw that I also had a `/opt/steam/libraryfolder.vdf`, so I just copied that contentid over to a second entry.

```
{
        "0"
        {
                "path"          "/home/uqs/.local/share/Steam"
                "label"         ""
                "contentid"             "1611104819401591390"
                "totalsize"             "0"
                "update_clean_bytes_tally"              "0"
                "time_last_update_corruption"           "0"
                "apps"
                {
                        "228980"                "883257504"
                        "1493710"               "1229396969"
                }
        }
        "1"
        {
                "path"          "/opt/steam"
                "label"         ""
                "contentid"             "2619151340883743646"
                "totalsize"             "0"
                "update_clean_bytes_tally"              "0"
                "time_last_update_corruption"           "0"
        }
}
```

Started steam, and voila presto it looks same as before. Though the "Storage" setting is weird, in that it shows Local Drive (/) _and_ External Drive (/)

So is it possible that my (not so old, mind you) installation of Steam with a separate games folder is no longer a supported setup? It sure is impossible to add it like so in the UI. It only seems to truly want to add different drives, not different locations from the same drive ...

**Comment by NathanielH-snek** [comment]
I think this isn't specific to Linux either. In addition to my Linux desktop I also have a mac and I occasionally run games off of an external drive. Sleeping the device while steam is still running will cause the external drive to "eject improperly" and steam will not recognize the drive anymore until manually readding the library. Admittedly this may be a separate issue but I wanted to contribute this in case it was useful

**Comment by TTimo** [comment]
Thanks @NathanielH-snek - this is clearly the same problem.

**Comment by micke1m** [comment]
I have the same issue occasionally on Arch. I believe it happens at OS shutdown if Steam is still running, the OS unmounts the drive or at least signals it is being removed before Steam shuts down. My steam library is on an internal SSD mounted at /ssd (/ssd/steamlibrary).
[content_log.txt](https://github.com/ValveSoftware/steam-for-linux/files/12103348/content_log.txt)

**Comment by NeoWordire** [comment]
same issue seen on manjaro. I have 3 internal ssd's beside a root ssd and all non root need to be manually re-added to the steam storage on occasion after shutdown. about once a week.

**Comment by TTimo** [comment]
This issue should be fixed in latest beta (Aug 18 build).

**Comment by zany130** [comment]
fixed for me

**Comment by LouieMiddle** [comment]
> This issue should be fixed in latest beta (Aug 18 build).

I'm still getting this issue on ubuntu 22.04.03

**Comment by sschuber** [comment]
For me the problme is still there on Linux Mint 21.2 Cinnamon with (deb) steam client:
Steam-Version:  1696019606
Steam-Client: Build-Datum:  Fr., 29. Sept. 19:38 UTC -08:00

**Comment by gardenboi** [comment]
Folder structure defaults for automatic recognition is `Disk`->`SteamLibrary`
this may solve the issue

**Comment by josiahcarlson** [comment]
Recently have experienced this problem, or a similar / related one in #10207.

```
$ cat /home/josiah/.local/share/Steam/config/libraryfolders.vdf
"libraryfolders"
{
	"0"
	{
		"path"		"/home/josiah/.local/share/Steam"
		"label"		""
		"contentid"		"5825306130931075617"
		"totalsize"		"0"
		"update_clean_bytes_tally"		"19153542857"
		"time_last_update_corruption"		"0"
		"apps"
		{
			"2610"		"2856881654"
			"105600"		"715107635"
			"212680"		"413405254"
			"228980"		"325850690"
			"233450"		"550031103"
			"244850"		"38908453485"
			"268130"		"199464047"
			"466300"		"1788376799"
			"588650"		"2097862719"
			"858280"		"375543351"
			"961940"		"399675883"
			"1054830"		"506437594"
			"1070560"		"12779"
			"1113280"		"910331459"
			"1321660"		"19131135224"
			"1391110"		"644161696"
			"1420170"		"909574564"
			"1493710"		"1234949917"
			"1887720"		"1098535663"
		}
	}
}
```
I don't know if this helps:
```
$ cat /mnt/bigpool/steam/libraryfolder.vdf
"libraryfolder"
{
	"contentid"		"7853347000011732665"
	"label"		""
}
```

[compat_log.previous.txt](https://github.com/ValveSoftware/steam-for-linux/files/13320725/compat_log.previous.txt)
[content_log.previous.txt](https://github.com/ValveSoftware/steam-for-linux/files/13320726/content_log.previous.txt)
[client_networkmanager.txt](https://github.com/ValveSoftware/steam-for-linux/files/13320751/client_networkmanager.txt)

**Comment by Jose123456** [comment]
I've started to experience this on Void Linux. Don't know exactly when it started but it was in the last couple of months. I didn't notice for a while, but I thought it was weird none of my games had updates.

I can add it back fine through the Settings->Storage dialog. There's a `libraryfolder.vdf` on the second drive, but none under `~/.local/share/Steam`.

**Comment by theSplund** [comment]
Getting the same problem on my Windows 10/Ubuntu/Mint machine. It works fine in Mint (Cinnamon), but Ubuntu 22, and now 23, simply forgets my 'other drives'

**Comment by Nikanoru** [comment]
I also have this problem on Ubuntu 22.04 LTS. 
I am using one 2TB m.2 drive which has 1 Parition for Windows NTFS, 1 Partition for Storage and my Steamlibrary NTFS, and 1 Partition for Ubuntu. Every time I restart my computer, Steam shows no entries in the library HOWEVER if I open a folder on my NTFS Partition that has my steam library on it (which causes the partition to be mounted), my steam library instantly fills up wit all the games.

So my uneducated guess would be to tell the OS to mount that partition on startup?

**Comment by theSplund** [comment]
> I also have this problem on Ubuntu 22.04 LTS. I am using one 2TB m.2 drive which has 1 Parition for Windows NTFS, 1 Partition for Storage and my Steamlibrary NTFS, and 1 Partition for Ubuntu. Every time I restart my computer, Steam shows no entries in the library HOWEVER if I open a folder on my NTFS Partition that has my steam library on it (which causes the partition to be mounted), my steam library instantly fills up wit all the games.
> 
> So my uneducated guess would be to tell the OS to mount that partition on startup?

I'm fairly uneducated (on Linux) too - I did try mounting at startup, but all I got was a bunch of mounted drives on my desktop and Steam was still missing the libraries I'd previously 'mapped'.
A heads up though: I eventually figured out that I really needed to install the games on Linux's Ext4 partition as only 6 of the 36 games that were on my NTFS partition wanted to work in Linux. As soon as I copied/installed a few of them into the Ext4 partition they started to work in Linux - I understand it to be a read/write issue (?) and you seem to have to commit to installing them under one OS or the other and boot to that particular OS (still, they did work).

**Comment by Jose123456** [comment]
I also had all kinds of weird problems when my Steam library was on an NTFS drive. I moved to ext4 and most games Just Work now.

**Comment by ezdev128** [comment]
It's happens just yesterday. Sadly, almost 2 years since this bug report...

**Comment by ezdev128** [comment]
The identical problem occurred today, but with a different sequence of events. 
I left Dota2 running overnight and in the morning, I realized that it had frozen. When I tried clicking the close button, there was no response. Then I've closed the entire Steam application and then Dota2 has closed, but in result in the secondary library being lost once again.

**Comment by hamletmillqvist** [comment]
Same problem on brand new installation of Ubuntu 24.04 LTS.
Secondary library is located on a partition in a separate drive. Separate drive is internally connected like the system drive.

Re-adding it every time works, but is annoying.

**Comment by Meek97** [comment]
I was running into this issue as well
OS: Linux Mint 21.3 Cinnamon
Steam: 1716242052 (package v 1.0.0.79 installed from .deb)

I think the root of my problem is that my secondary drive wasn't actually be automatically mounted at startup.

Under `Mount Options` for the drive in question, `User Session Defaults` was enabled, leaving all other options grayed out but `Mount at system startup` was checked. This must not have been the `session default` because i found out it was in fact not being mounted at start-up. If i unchecked `User Session Defaults` and made sure to have `Mount at system startup` checked, I confirmed that my secondary drive is now being mounted at start-up, and steam is no longer forgetting about my secondary drive :)

Mount options with `User Session Defaults` enabled -> this drive was NOT being mounted at start with this config
![image](https://github.com/ValveSoftware/steam-for-linux/assets/14891222/03eadd88-ea9a-4d83-b4be-eef62358add9)

Mount options with `User Session Defaults` disabled -> now my drive mounts at startup
![image](https://github.com/ValveSoftware/steam-for-linux/assets/14891222/15c16127-878f-40a3-9a59-509bddf1a043)

**Comment by circuitsystem** [comment]
this has been happening to me ever since i installed mint for the very first time a week ago. Meek97's solution of auto-mounting on "disks" doesnt fix the problem

**Comment by 43615** [comment]
I have encountered this due to my usage of Satisfactory's dedicated server tool. If I use external methods of shutting it down (often useful due to lack of a GUI), my main library directory will be missing after Steam restarts.
- Up-to-date Arch Linux
- Steam version 1726256783
- The directory is a secondary drive, persistently mounted via `/etc/fstab`

**Comment by Zsar** [comment]
Happening to me on Arch using the Arch-provided Steam package since ~2025-08-01. Has never happened before, since then it happens 100% of the time.

(I am not entirely sure of the date, as I did not have Steam open every single day.)

My first library is '/' (actual path `~/.local/share/Steam/`. My second library is '/mnt/ssd0' (actual path `/mnt/ssd0/SteamLibrary/`) and marked as default.

Whenever I start Steam, the second library is gone and the default transferred to the first library. Even if I start Steam immediately after closing it.

I always close Steam using its menu bar entry `Steam|Exit`.

Steam package:
```
$ pacman -Qii steam
Name            : steam
Version         : 1.0.0.82-2
Description     : Valve's digital software delivery system
Architecture    : x86_64
URL             : https://steampowered.com/
Licenses        : LicenseRef-steam-subscriber-agreement
Groups          : None
Provides        : None
Depends On      : bash  coreutils  curl  dbus  desktop-file-utils  diffutils  freetype2  gcc-libs  gdk-pixbuf2  glibc  hicolor-icon-theme  libxcrypt  libxcrypt-compat
                  libxkbcommon-x11  lsb-release  lsof  nss  python  ttf-font  usbutils  vulkan-driver  vulkan-icd-loader  xdg-user-dirs  xorg-xrandr  xz  zenity
                  lib32-alsa-plugins  lib32-fontconfig  lib32-gcc-libs  lib32-glibc  lib32-libgl  lib32-libgpg-error  lib32-libnm  lib32-libva  lib32-libx11  lib32-libxcrypt
                  lib32-libxcrypt-compat  lib32-libxinerama  lib32-libxss  lib32-nss  lib32-pipewire  lib32-systemd  lib32-vulkan-driver  lib32-vulkan-icd-loader
Optional Deps   : polkit: to setup SteamVR without root access [installed]
                  steam-native-runtime: steam native runtime support [installed]
                  xdg-desktop-portal-impl: file & folder picker [installed]
Required By     : steam-native-runtime
Optional For    : None
Conflicts With  : None
Replaces        : None
Installed Size  : 18.65 MiB
Packager        : Levente Polyak <anthraxx@archlinux.org>
Build Date      : Tue 01 Apr 2025 17:37:22 CEST
Install Date    : Sun 06 Apr 2025 14:54:05 CEST
Install Reason  : Explicitly installed
Install Script  : No
Validated By    : Signature
Backup Files    : None
Extended Data   : pkgtype=pkg
```

Mount point:
```
$ cat /etc/fstab 
# /dev/sda1
UUID=af0d636f-d2fb-441a-b03f-0b70a9186ba3	/         	ext4      	rw,relatime	0 1

# /dev/nvme0n1p1
UUID=0459-BCFA      	/boot     	vfat      	rw,relatime,fmask=0022,dmask=0022,codepage=437,iocharset=ascii,shortname=mixed,utf8,errors=remount-ro	0 2

# /dev/nvme0n1p2
UUID=523df5c3-6c01-48ef-87c7-97f54242eab4	/mnt/ssd0	ext4	rw,relatime,discard	0	2

# /dev/nvme0n1p2 swapfile
/mnt/ssd0/swapfile	none	swap	defaults	0	0
```
